Rewrite ACM translator for new site

GuyAglionby opened the pull request:
Supports the new ACM site (#1811), which has changed over already. The tests don't run (see below), but I've manually tested that it all works. 

A couple of notes: I return that magazines and newsletters are `journalArticle` as this is the type exported by their CSL. I also wasn't sure what the best way of dealing with ISBN was in cases where it is provided but the appropriate item type doesn't have a field for it, so put it in the extras. 

With regard to the conversation in #2097, a conference proceedings page will bring up a dialog for users to select papers in that proceedings rather than download the proceedings information. This is easily changed depending on what we decide to be most reasonable. 

There is a slight problem with the new site in that it does not work without the cookie `cookiePolicy=accept` set: it just redirects to https://dl.acm.org/action/cookieAbsent. As a result I've not been able to run the translator in Scaffold, either manually or with the tests. Manually setting the headers in `doGet` to include the cookie doesn't work either, as it looks like XMLHttpRequest doesn't allow that. In any case I don't think that would address the problem with the tests.

zuphilip commented on ACM Digital Library.js:
```diff
@@ -1,203 +1,200 @@
+		let abstractElements = doc.querySelectorAll('div.article__abstract p, div.abstractSection p');
+		let abstract = Array.from(abstractElements).map(function (element) {
+			return element.textContent;
+		}).join(' ');
```
You can also use the short form
```
map(x => x.textContent)
```
Moreover, I would suggest to use `.join('\n\n')` instead, because you are dealing here with different paragraphs in `p` nodes.

GuyAglionby replied:
I forgot about this notation, thanks

zuphilip commented on ACM Digital Library.js:
```diff
@@ -1,203 +1,200 @@
+			return 'computerProgram';
+		}
+		else if (subtype == 'dataset') {
+			return 'document';
```
Add in this case then also `itemType: data` in the extra somewhere below.

zuphilip commented on ACM Digital Library.js:
```diff
@@ -1,203 +1,200 @@
+			// e.g. https://dl.acm.org/doi/abs/10.5555/3336323.C5474411
+			let authorElements = doc.querySelectorAll('div.citation span.loa__author-name');
+			authorElements.forEach(function (element) {
+				item.creators.push(ZU.cleanAuthor(element.textContent));
```
:bug: This function `cleanAuthor` needs a second argument for the creatorType!

zuphilip commented on ACM Digital Library.js:
```diff
@@ -1,203 +1,200 @@
+			// A chapter of a book is extracted as a bookSection, but the
+			// authors are not included in the CSL so we must scrape them.
+			// e.g. https://dl.acm.org/doi/abs/10.5555/3336323.C5474411
+			let authorElements = doc.querySelectorAll('div.citation span.loa__author-name');
```
We could do this always when `item.creators` array is empty but `authorElements` not, i.e. there is no reason to restrict this to bookSections only.

zuphilip commented on ACM Digital Library.js:
```diff
@@ -1,203 +1,200 @@
+		}
+		
+		let pagesElement = doc.querySelector('div.pages-info span');
+		if (pagesElement) {
```
```suggestion
		if (pagesElement && !item.numPages) {
```

zuphilip commented on ACM Digital Library.js:
```diff
@@ -1,203 +1,200 @@
-		return getSearchResults(doc, true) ? 'multiple' : false;
-	} else if (url.includes("/citation.cfm")) {
-		return getArticleType(doc);
+	if ((url.includes('/doi/') || url.includes('/do/')) && !url.includes('/doi/proceedings')) {
```
Can you also give an example/test case for a `/do/` url?

GuyAglionby replied:
`/do/` includes [computer programs](https://dl.acm.org/do/10.1145/3159287/full/) and [datasets](https://dl.acm.org/do/10.1145/3276998/full/). However, it looks like the CSL endpoint currently gives a 500 when trying to retrieve information for the DOIs corresponding to these types of entry (I checked a few). This also breaks the 'export citation' functionality on the website itself, though I recall it working before. Not sure if it's worth disabling `/do/` in the mean time (I'm not sure how long it's been broken, but have just emailed about it).

zuphilip replied:
Thank you for the examples. No, we don't need to disable this here. As far as we know this should work on the Zotero side. Thank you also for reaching out to them (metadata matters!).

zuphilip commented:
@GuyAglionby Let me know if you can work on this within the next days. Otherwise, I can maybe try to work a little further here.

GuyAglionby commented:
Sorry about the delay here -- I should have time tomorrow to have a look

GuyAglionby commented on ACM Digital Library.js:
```diff
@@ -1,203 +1,200 @@
+	GNU Affero General Public License for more details.
 
-// attr()/text() v2
-function attr(docOrElem,selector,attr,index){var elem=index?docOrElem.querySelectorAll(selector).item(index):docOrElem.querySelector(selector);return elem?elem.getAttribute(attr):null;}function text(docOrElem,selector,index){var elem=index?docOrElem.querySelectorAll(selector).item(index):docOrElem.querySelector(selector);return elem?elem.textContent:null;}
```
I misremembered the discussion in another issue re. those functions as getting rid of them entirely rather than incorporating them as globals. Thanks for the heads up!

zuphilip replied:
No problem and you remember correctly, that this is the goal. However, AFAIK we are not yet there to drop these lines. Anyways it should be easy to do this then for a lot of translators at once, when it is ready.

GuyAglionby commented on ACM Digital Library.js:
```diff
@@ -1,203 +1,200 @@
+		if (subtype == 'conference') {
+			return 'conferencePaper';
+		}
+		else if (subtype == 'journal' || subtype == 'magazine' || subtype == 'newsletter') {
```
[Here](https://dl.acm.org/doi/10.1145/2854146) is a magazine ([corresponding test](https://github.com/zotero/translators/pull/2101/commits/a853a19e0dfa55c877a5731e71378e362f9f00a1#diff-2b2fb3a06b538f7aa0f0d1f03670c891R558)); [here](https://dl.acm.org/doi/10.1145/3264631.3264634) is a newsletter ([corresponding test](https://github.com/zotero/translators/pull/2101/commits/a853a19e0dfa55c877a5731e71378e362f9f00a1#diff-2b2fb3a06b538f7aa0f0d1f03670c891R518)).

zuphilip replied:
So, they don't have an submit area for authors and are maybe otherwise a little different. However, I see that dblp is classify them also as journals. Thus, I am okay with this now.

GuyAglionby commented on ACM Digital Library.js:
```diff
@@ -1,203 +1,200 @@
 		
-		if (checkOnly) return true;
-		found = true;
+		if (!(url.includes('/doi/') || url.includes('/do/')) || url.includes('/doi/proceedings')) {
```
I think this was already a negation, but to avoid any future confusion I've pulled this check out into its own function

zuphilip replied:
Ah, the parentheses ...

GuyAglionby commented:
Thanks for the comments -- have made changes as requested. The tests started working for me in Scaffold (without any changes, not sure what happened) so I updated the already-present ones, and added new ones to cover both new item types and some of the additional scraping that is done.

One thing `ZU.itemFromCSLJSON` does is return creators with a `creatorTypeID`, which doesn't seem necessary although I'm not sure it's doing any harm. Can remove that field if necessary.

zuphilip commented:
Okay, great! Thank you very much! I have only some small nits pushed directly to your branch. I was not able to use Scaffold for testing. The additional field for the creators should be okay.

zuphilip commented:
Merged now! :rocket: Thank you very much! :bowing_man:

GuyAglionby commented:
Thanks!
